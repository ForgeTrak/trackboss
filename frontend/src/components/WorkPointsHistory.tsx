/* eslint-disable no-unused-vars */
/* eslint-disable react/no-unused-prop-types */
import _ from 'lodash';
import moment from 'moment';
import {
    Heading,
    VStack,
} from '@chakra-ui/react';
import React, { createRef, RefObject, useContext, useEffect, useState } from 'react';
import DataTable from 'react-data-table-component';
import { BsChevronDown } from 'react-icons/bs';
import { Job } from '../../../src/typedefs/job';
import { UserContext } from '../contexts/UserContext';
import { getJobList } from '../controller/job';
import { getWorkPointsByMembership, getWorkPointsTotal } from '../controller/workPoints';
import { getYearlyThreshold, getYearlyThresholdValue } from '../controller/billing';
// eslint-disable-next-line import/no-named-as-default
import AddPointsModal from './AddPointsModal';
import YearsDropDown from './shared/YearsDropDown';

const columns: any = [
    {
        name: 'Date',
        selector: (row: Job) => `${moment(row.start).utc().format('MM/DD/YYYY')}`,
        sortable: true,
        wrap: true,
    },
    {
        name: 'Job',
        selector: (row: Job) => `${row.title}`,
        wrap: true,
        sortable: true,
    },
    {
        name: 'Points',
        selector: (row: Job) => {
            const isFutureSignup = moment(row.end).isAfter();
            if (isFutureSignup) {
                return `(${row.pointsAwarded} pending)`;
            }
            if (row.paid) {
                return 0;
            }
            return row.pointsAwarded;
        },
        sortable: true,
    },
    {
        name: 'Earned By',
        selector: (row: Job) => `${row.member}`,
        wrap: true,
        sortable: true,
    },
];

const customStyles = {
    headCells: {
        style: {
            paddingTop: '0',
            fontSize: '1.5em',
            backgroundColor: '#f9f9f9',
            color: '#626262',
        },
    },
    cells: {
        style: {
            fontSize: '1.2em',
        },
    },
};

export default function WorkPointsHistory() {
    const [cells, setCells] = useState<Job[]>([]);
    const [allJobs, setAllJobs] = useState<Job[]>([]);
    const [year, setYear] = useState<number>(new Date().getFullYear());
    const [years, setYears] = useState<number[]>([]);
    const [workPointsEarned, setWorkPointsEarned] = useState<number>(0);
    const [workPointsThreshold, setWorkPointsThreshold] = useState<number>(0);

    const { state } = useContext(UserContext);

    async function getJobsData() {
        const jobs =
            await getJobList(state.token, 'membershipID', state.user!.membershipId as unknown as string) as Job[];
        jobs.sort((a, b) => {
            let order = 0;
            if (a.start < b.start) {
                order = 1;
            } else if (a.start > b.start) {
                order = -1;
            }
            return order;
        });
        if ('reason' in jobs) {
            // squash error
        } else {
            const allYears: number[] = [];
            // find the first year they were a member.
            let firstYear = parseInt(state.user?.dateJoined.substring(0, 4) || '0', 10);
            // but we tracked points offline before 2016 so we can only go from then on.
            if (firstYear < 2016) {
                firstYear = 2016;
            }
            const currentYear = new Date().getFullYear();
            for (let dropdownYear = firstYear; dropdownYear <= currentYear; dropdownYear++) {
                allYears.push(dropdownYear);
            }
            // reverse sort the years so that most recent is always at the top
            const sortedYears = allYears.sort((a, b) => (b - a));
            setYears(sortedYears);
            setAllJobs(jobs);
        }
    }

    useEffect(() => {
        getJobsData();
    }, []);

    useEffect(() => {
        setCells(_.filter(allJobs, (job) => new Date(job.start).getFullYear() === year));
        async function getData() {
            const workPoints = await getWorkPointsByMembership(state.token, state.user!.membershipId, year);
            const threshold = await getYearlyThreshold(state.token, year);
            if ('reason' in workPoints) {
                setWorkPointsEarned(0);
            } else {
                setWorkPointsEarned(workPoints.total);
            }
            if ('reason' in threshold) {
                // squash
            } else {
                setWorkPointsThreshold(threshold.threshold);
            }
        }
        getData();
    }, [year, allJobs]);

    const userActive = state.user?.active || false;
    return (
        <VStack bg="white">
            <YearsDropDown
                years={years}
                header="Work Points History"
                setYear={setYear}
                initialYear={(new Date()).getFullYear()}
            />
            <Heading color="orange" size="2xl">
                {`${workPointsEarned}/${workPointsThreshold}`}
            </Heading>
            <AddPointsModal
                memberName={state?.user?.firstName || ''}
                membershipId={state?.user?.membershipId || 0}
                memberId={state?.user?.memberId as number}
                // eslint-disable-next-line max-len
                visible={((state.storedUser?.memberType === 'Admin' || state.user?.memberType === 'Admin') && (userActive))}
                token={state.token}
                // eslint-disable-next-line react/jsx-no-bind
                refreshPoints={getJobsData}
            />
            <DataTable
                columns={columns}
                data={cells}
                fixedHeaderScrollHeight="300px"
                highlightOnHover
                pagination
                responsive
                subHeaderWrap
                customStyles={customStyles}
            />
        </VStack>
    );
}
