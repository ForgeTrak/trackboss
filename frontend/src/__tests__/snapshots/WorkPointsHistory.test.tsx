import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import WorkPointsHistory from '../../components/WorkPointsHistory';

describe('work points history', () => {
    it('renders all props correctly', () => {
        const header = renderer.create(
            <BrowserRouter>
                <WorkPointsHistory />
            </BrowserRouter>,
        ).toJSON();
        expect(header).toMatchSnapshot();
    });
});
