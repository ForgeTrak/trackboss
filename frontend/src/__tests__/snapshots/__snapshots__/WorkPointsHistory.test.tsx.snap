// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`work points history renders all props correctly 1`] = `
<div
  className="chakra-stack css-yth84b"
>
  <div
    className="chakra-stack css-84zodg"
  >
    <button
      aria-controls="menu-list-1"
      aria-expanded={false}
      aria-haspopup="menu"
      className="chakra-button chakra-menu__menu-button css-1bur0ct"
      id="menu-button-1"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      <span
        className="css-xl71ch"
      >
        Past Years
      </span>
      <span
        className="chakra-button__icon css-1hzyiq5"
      >
        <svg
          aria-hidden={true}
          fill="currentColor"
          focusable={false}
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
    <div
      className="css-0"
      style={
        Object {
          "inset": "0 auto auto 0",
          "minWidth": "max-content",
          "position": "absolute",
          "visibility": "hidden",
        }
      }
    >
      <div
        aria-orientation="vertical"
        className="chakra-menu__menu-list css-sjt5nk"
        id="menu-list-1"
        onKeyDown={[Function]}
        role="menu"
        style={
          Object {
            "opacity": 0,
            "transform": "scale(0.8) translateZ(0)",
            "transformOrigin": "var(--popper-transform-origin)",
            "visibility": "hidden",
          }
        }
        tabIndex={-1}
      />
    </div>
    <h2
      className="chakra-heading css-0"
    >
      Work Points History
    </h2>
  </div>
  <h2
    className="chakra-heading css-7kmldq"
  >
    0/0
  </h2>
  <div
    className="sc-dlVxhl fSQziN"
  >
    <div
      className="sc-fKVqWL XlRpR"
    >
      <div
        className="sc-bdvvtL hLGqlg rdt_Table"
        disabled={false}
        role="table"
      >
        <div
          className="sc-iwjdpV kntihR"
        >
          <div
            style={
              Object {
                "padding": "24px",
              }
            }
          >
            There are no records to display
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
